lib/theme_check/node.rb

```diff
@@ -101,19 +112,38 @@ def type_name
       @type_name ||= StringHelpers.underscore(StringHelpers.demodulize(@value.class.name)).to_sym
     end
 
+    def source
+      template&.source
+    end
+
     # Is this node inside a `{% liquid ... %}` block?
     def inside_liquid_tag?
-      if line_number
-        template.excerpt(line_number).start_with?("{%")
+      @inside_liquid_tag ||= if tag? && line_number && source
+        i = 1
+        i += 1 while source[start_index - i] =~ /\s/ && i < start_index
+        !(source[(start_index - i - 1)..(start_index - i)] =~ /%-?$/)
```

Comment:
Perhaps just a comment on top to explain?

That might help a tiny bit to simplify the logic (refactoring the negation to `=~` => `!~`):

```suggestion
        source[(start_index - i - 1)..(start_index - i)] !~ /%-?$/
```

Also, should you be matching the `{` too?

Reply:
> Also, should you be matching the { too?

If I did it would make the code harder to understand since I'd have to go back 3 characters and that might make me go to -1 in the following scenario:

```
{% assign x = 1 %}
```

Since I'd have:
```
start_index == 3
i = 2
start_index - i - 2 == -1
source[(start_index - i - 2)..(start_index - i)]
== source[-1..1]
== ""
```